snuba/datasets/errors_replacer.py

```diff
@@ -298,8 +300,57 @@ def _build_group_replacement(
     )
 
 
+def _build_event_set_filter(
```

Comment:
How much refactoring of this class are you willing to do :) ? (None is a good answer as well as that is understandably not in the scope of the project you are working on).

I am asking because the design of this class has outlived its original purpose and building new feature into this without changing the structure is becoming increasingly confusing. 
Specifically my problem is with the idea of producing string templates for the queries that are filled in with untyped and unstructured argument bags (query_args) that are partially populated here and in partially in replacer.py. This makes it confusing as there is a tight coupling between the parameters and the template, implying there will be a tight coupling with this class and the callsite.
When we add more indirection (like this method) we highlight that issue even more.

Now, if you are up for making a change like this, again, not blocking on this as it is out of the scope of your project and we would deal with this in our backlog anyway. The change below should be in its own PR.
What about Having a base abstract class for the Replacement (instead of the data class we have today) that provides an interface like this:

```
class Replacement(ABC):
    def get_count_query(self, table: string) -> string

    def get_insert_query(self, table: string) -> string
```

Each replacement type has its own implementation ( that now can be unit tested) and encapsulates the logic to build the query without having to return a template.  replacer.py would just receive one object of a subclass of this, and just run the query without caring of what it does. 

In the future we could add abstraction on top of the query itself so that we return an AST that can be formatted by the processor.py. But I would definitively not go into that without breaking this class apart first.

Moreover the subclasses of Replacement would not have to deal with the replacement message as a bare dictionary. errors_replacer.py could be the one parsing the message and make sense of their structure, while the sublcasses of Replacement would be instantiated with the specific fields they need and never see the dictionary.

What do you think ?

Reply:
this seems fine, I will also likely not get around doing some kind of refactor in Q2 as there logical complexity will sharply increase. about the details of the new design, I'll have to try it out and see what sticks